corehq/apps/couch_sql_migration/couchsqlmigration.py

```diff
@@ -194,44 +219,167 @@ def _rebuild_queues(self, pool):
 
-    def _migrate_form_and_associated_models(self, couch_form):
-        sql_form = _migrate_form(self.domain, couch_form)
-        _migrate_form_attachments(sql_form, couch_form)
-        _migrate_form_operations(sql_form, couch_form)
+    def _migrate_form_and_associated_models(self, src_couch_form):
+
+        dst_couch_form = self._map_form_ids(src_couch_form)
+        sql_form = _migrate_form(dst_couch_form, self.dst_domain)
+        _migrate_form_attachments(self.src_domain, sql_form, dst_couch_form)
+        _migrate_form_operations(sql_form, dst_couch_form)
 
-        self._save_diffs(couch_form, sql_form)
+        self._save_diffs(dst_couch_form, sql_form)
 
         case_stock_result = None
         if sql_form.initial_processing_complete:
-            case_stock_result = _get_case_and_ledger_updates(self.domain, sql_form)
+            case_stock_result = _get_case_and_ledger_updates(self.dst_domain, sql_form)
             if len(case_stock_result.case_models):
                 touch_updates = [
-                    update for update in get_case_updates(couch_form)
+                    update for update in get_case_updates(dst_couch_form)
                     if len(update.actions) == 1 and isinstance(update.actions[0], CaseNoopAction)
                 ]
                 if len(touch_updates):
                     # record these for later use when filtering case diffs. See ``_filter_forms_touch_case``
-                    self.forms_that_touch_cases_without_actions.add(couch_form.form_id)
+                    self.forms_that_touch_cases_without_actions.add(dst_couch_form.form_id)
 
         _save_migrated_models(sql_form, case_stock_result)
 
+    def _build_id_map(self):
+        """
+        Iterate locations and mobile workers to map IDs in the source
+        domain to IDs in the destination domain.
+
+        These will be used to update forms as they are migrated.
+
+        Apps and app builds are not mapped because they are not all in
+        the destination domain.
+        """
+        self._id_map = {}
+        for location in SQLLocation.objects.filter(domain=self.dst_domain):
+            if 'orig_id' in location.metadata:
+                self._id_map[location.metadata['orig_id']] = location.location_id
+
+        for user in CommCareUser.by_domain(self.dst_domain):
+            if 'orig_id' in user.user_data:
+                self._id_map[user.user_data['orig_id']] = user.get_id
+
+    def _map_form_ids(self, couch_form):
+        """
+        The destination domain will have new IDs for locations, mobile
+        workers, apps and app builds. Return the given form with source
+        domain IDs mapped to the corresponding IDs in the destination
+        domain.
+        """
+        if self.src_domain == self.dst_domain:
+            return couch_form
+
+        # Leave app_id
+        # Leave build_id
+
+        couch_form.domain = self.dst_domain
+
+        if 'case' in couch_form.form:
```

Comment:
I think you could leverage the existing case processing code we have like [extract_case_blocks](https://github.com/dimagi/commcare-hq/blob/8825b460cd87d96ce9e5f8e6c69cbc3d87d5b8dd/corehq/ex-submodules/casexml/apps/case/xform.py#L324) and [get_case_updates](https://github.com/dimagi/commcare-hq/blob/8825b460cd87d96ce9e5f8e6c69cbc3d87d5b8dd/corehq/ex-submodules/casexml/apps/case/xform.py#L388)